### PR TITLE
test/librbd: add unit tests for rbd update features

### PR DESCRIPTION
**Pull request by ideepika**
this additon is just for added coverage which I was using to debug/reproduce
rbd upgrade tests failures.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/test/librbd/test_ImageWatcher.cc [hunk 1]
```diff
@@ -214,6 +214,15 @@ class TestImageWatcher : public TestFixture {
       {
         RebuildObjectMapPayload payload;
         payload.decode(2, iter);
+        cout << payload.async_request_id;
```
Comment (tchaikov) [inline src/test/librbd/test_ImageWatcher.cc:217]: i'd avoid writing to `cout` in tests, unless something unexpected happens.

File: src/librbd/ImageWatcher.cc [hunk 2]
```diff
@@ -1382,6 +1384,8 @@ void ImageWatcher<I>::process_payload(uint64_t notify_id, uint64_t handle,
   auto ctx = new Watcher::C_NotifyAck(this, notify_id, handle);
   bool complete;
 
+  ldout(cct, 10) << this << " process_payload "  << payload->get_notify_op() << dendl;
```
Comment (trociny) [inline src/librbd/ImageWatcher.cc:1387]: I am not sure how useful it is. We have already a message in `handle_payload`, which is called just below and usually provides much more information. In what case do you think it is useful?

File: src/test/librbd/test_ImageWatcher.cc (hunk 1 above)
Comment (trociny) [reply src/test/librbd/test_ImageWatcher.cc:217]: Yes. Historically it is not always true for all our unit tests, but I would not add it in the new ones. You can get this info running the test with CEPH_ARGS=--debug_rbd=20 from the ceph client log.

**Review by trociny — COMMENTED** [review]
otherwise LGTM

File: src/librbd/ImageWatcher.cc (hunk 2 above)
Comment (ideepika) [reply src/librbd/ImageWatcher.cc:1387]: I used this to identify what payload { AsyncComplete or UpdateFeatures was being processed and to judge by it has the requested later or not, maybe we can add just ` payload->get_notify_op() ` to handle_payload info itself

File: src/librbd/image_watcher/NotifyLockOwner.cc [hunk 3]
```diff
@@ -80,6 +80,7 @@ void NotifyLockOwner::handle_notify(int r) {
     decode(response_message, iter);
 
     r = response_message.result;
+    ldout(cct, 1) << " client responded with r=" << r << dendl;
```
Comment (trociny) [inline src/librbd/image_watcher/NotifyLockOwner.cc:83]: nit: I think the log level 1 is too low for a message like this. I would use 10 or 20.

File: src/test/librbd/test_support.h [hunk 4]
```diff
@@ -28,11 +28,11 @@ bool is_librados_test_stub(librados::Rados &rados);
...
 bool is_rbd_pwl_enabled(ceph::common::CephContext *ctx);
 
-#define REQUIRE(x) {			  \
-  if (!(x)) {				  \
-    std::cout << "SKIPPING" << std::endl; \
-    return SUCCEED(); 			  \
-  } 					  \
+#define REQUIRE(x) {							      \
+  if (!(x)) {								      \
+    std::cout << "SKIPPING TEST(REQUIRED FEATURE NOT ENABLED)" << std::endl; \
```
Comment (trociny) [inline src/test/librbd/test_support.h:33]: I am not quite sure I like this. It looks too verbose as for a standard test message and it is used not only for features but also for config options.

File: src/librbd/image_watcher/NotifyLockOwner.cc (hunk 3 above)
Comment (ideepika) [reply src/librbd/image_watcher/NotifyLockOwner.cc:83]: makes sense +1

**Review by trociny — APPROVED** [review]
:+1:

**Comment by yuriw** [comment]
@trociny see https://trello.com/c/7JVOzvqo https://pulpito.ceph.com/yuriw-2021-08-12_20:49:44-rbd-wip-yuri-testing-2021-08-12-1127-distro-basic-smithi

**Comment by trociny** [comment]
> @trociny see https://trello.com/c/7JVOzvqo https://pulpito.ceph.com/yuriw-2021-08-12_20:49:44-rbd-wip-yuri-testing-2021-08-12-1127-distro-basic-smithi

The test failures look not related to this PR.

**Comment by ideepika** [comment]
> > @trociny see https://trello.com/c/7JVOzvqo https://pulpito.ceph.com/yuriw-2021-08-12_20:49:44-rbd-wip-yuri-testing-2021-08-12-1127-distro-basic-smithi
> 
> The test failures look not related to this PR.

@trociny good to merge?

**Pull request merged**